File: Algorithms/Flood_Fill.py
'''


The idea is simple. Simply perform a DFS on the source cell. Continue the DFS if:

Next cell is within bounds.
Next cell is the same color as source cell.
There is a tricky case where the new color is the same as the original color and if the DFS is done on it, 
there will be an infinite loop. If new color is same as original color, 
there is nothing to be done and we can simply return the image.



'''


class Solution(object):
    def floodFill(self, image, sr, sc, newColor):
        rows, cols, orig_color = len(image), len(image[0]), image[sr][sc]
        def traverse(row, col):
            if (not (0 <= row < rows and 0 <= col < cols)) or image[row][col] != orig_color:
                return
            image[row][col] = newColor
            [traverse(row + x, col + y) for (x, y) in ((0, 1), (1, 0), (0, -1), (-1, 0))]
        if orig_color != newColor:
            traverse(sr, sc)
        return image
      
      
      
################## Another solution  ##################


class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        r, c = len(image), len(image[0])
        color = image[sr][sc]
        def dfs(i, j):
            if i < 0 or i>=r or j < 0 or j >= c:
                return
            if image[i][j] == newColor or image[i][j] != color:
                return
            image[i][j] = newColor
            dfs(i+1, j)
            dfs(i-1, j)
            dfs(i,j+1)
            dfs(i, j-1)
        dfs(sr, sc)
        return image
        
        
        

#################       BFS        ###########################


class Solution:
    def floodFill(self, image, sr, sc, newColor):
        old, m, n = image[sr][sc], len(image), len(image[0])
        if old != newColor: 
            q = collections.deque([(sr, sc)])
            while q:
                i, j = q.popleft()
                image[i][j] = newColor
                for x, y in ((i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)):
                    if 0 <= x < m and 0 <= y < n and image[x][y] == old: 
                        q.append((x, y))
        return image
      
      
      
####################      DFS          ###########################


class Solution:
    def floodFill(self, image, sr, sc, newColor):
        def dfs(i, j):
            image[i][j] = newColor
            for x, y in ((i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)):
                if 0 <= x < m and 0 <= y < n and image[x][y] == old:
                    dfs(x, y)
        old, m, n = image[sr][sc], len(image), len(image[0])
        if old != newColor: 
            dfs(sr, sc)
        return image
      
      
      
      
############## Minimum liners #####################


def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
    """ O(N*M)TS in place """

    def fn(y, x, color):
        if 0 <= y < len(image) and 0 <= x < len(image[0]) and image[y][x] == color:
            image[y][x] = newColor
            fn(y + 1, x, color), fn(y - 1, x, color), fn(y, x + 1, color), fn(y, x - 1, color)

    return (fn(sr, sc, image[sr][sc]) or image) if image[sr][sc] != newColor else image

  
  
  
#### Another solution



class Solution(object):
    def floodFill(self, image, sr, sc, newColor):
        R, C = len(image), len(image[0])
        color = image[sr][sc]
        if color == newColor: return image
        def dfs(r, c):
            if image[r][c] == color:
                image[r][c] = newColor
                if r >= 1: dfs(r-1, c)
                if r+1 < R: dfs(r+1, c)
                if c >= 1: dfs(r, c-1)
                if c+1 < C: dfs(r, c+1)

        dfs(sr, sc)
        return image
      

File: Algorithms/Flood_Fill.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if (image[sr][sc] == newColor) return image;
        fill(image, sr, sc, image[sr][sc], newColor);
        return image;
    }
    
    private void fill(int[][] image, int sr, int sc, int color, int newColor) {
        if (sr < 0 || sr >= image.length || sc < 0 || sc >= image[0].length || image[sr][sc] != color) return;
        image[sr][sc] = newColor;
        fill(image, sr + 1, sc, color, newColor);
        fill(image, sr - 1, sc, color, newColor);
        fill(image, sr, sc + 1, color, newColor);
        fill(image, sr, sc - 1, color, newColor);
    }
}


// Another solution

// BFS


class Solution {
    int[] dr = new int[] {0, 0 , 1, -1};
    int[] dc = new int[] {-1, 1, 0, 0};
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        int R = image.length;
        int C = image[0].length;
        
        int oldColor = image[sr][sc];
        if (oldColor == newColor)
            return image;
        Queue<Point> q = new ArrayDeque<Point>();
        q.add(new Point(sr, sc));
        
        while(!q.isEmpty()) {
            Point p = q.poll();
            image[p.x][p.y] = newColor;
            for (int i = 0; i < 4; i++) {
                int x = p.x + dr[i];
                int y = p.y + dc[i];
                if (isValid(image, oldColor, x, y, R, C)) {
                    q.add(new Point(x, y));
                }
            }
        }
        return image;
    }
    
    
     private boolean isValid(int[][] image, int oldColor, int x, int y, int R, int C) {
            if (x >= 0 && x < R && y >= 0 && y < C && image[x][y] == oldColor)
                return true;
         return false;
        }
    
    class Point{
        int x;
        int y;
        
        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
        
    }
}



// DFS


class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        int o=image[sr][sc];
        helper(image,sr,sc,newColor,o);
        return image;

    }

    void helper(int[][] img,int r,int c,int n,int o) {
        if(r>=img.length || c>=img[0].length || r<0 || c<0 || img[r][c]==n || img[r][c]!=o)   return;
        img[r][c]=n;
        helper(img,r+1,c,n,o);
        helper(img,r-1,c,n,o);
        helper(img,r,c+1,n,o);
        helper(img,r,c-1,n,o);
    }
}




// Another solution


class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if(image[sr][sc] == newColor)
            return image;
        floodFill(image,sr,sc,image[sr][sc],image.length,image[0].length,newColor);
        return image;
    }

    public void floodFill(int[][] image,int cr,int cc,int color,int m,int n,int newColor){

        if(image[cr][cc] != color)
            return;

        image[cr][cc] = newColor;
        if(cr >= 1)
            floodFill(image,cr-1,cc,color,m,n,newColor);
        if(cc < n-1)
            floodFill(image,cr,cc+1,color,m,n,newColor);
        if(cr < m-1)
            floodFill(image,cr+1,cc,color,m,n,newColor);
        if(cc >= 1)
            floodFill(image,cr,cc-1,color,m,n,newColor);
    }
}




// Another



class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        int color = image[sr][sc];
        if (color != newColor) dfs(image, sr, sc, color, newColor);
        return image;
    }
    public void dfs(int[][] image, int r, int c, int color, int newColor) {
        if (image[r][c] == color) {
            image[r][c] = newColor;
            if (r >= 1) dfs(image, r-1, c, color, newColor);
            if (c >= 1) dfs(image, r, c-1, color, newColor);
            if (r+1 < image.length) dfs(image, r+1, c, color, newColor);
            if (c+1 < image[0].length) dfs(image, r, c+1, color, newColor);
        }
    }
}
